File: examples/Python/request_image.py
#! /usr/bin/env python
from utils import *
import base64


def main():
    status = parse_json(stdin.readline())

    # we start checking if the channel is ready
    if status['status'] != 'ready':
        raise Exception("There is problem in the communication channel")

    write_line('''
        {
            "type": "spider",
            "name": "image",
            "start_urls": []
        }
        ''')

    write_line('''
            {
                "type": "request",
                "id": "file",
                "url": "http://httpbin.org/image/png",
                "base64": true
            }''')

    msg = parse_json(stdin.readline())

    # check the message type
    if msg['type'] == 'exception':
        raise Exception("Failed to get content")
    elif msg['type'] == 'response':

        img = base64.b64decode(msg['body'])
        with open('outputs/image.png', 'wb') as f:
            f.write(img)
        write_line('{"type": "log", "level": "debug", "message": "DONE"}')

    write_line('{"type": "close"}')

if __name__ == '__main__':
    main()

File: examples/Java/RequestImage.java
import org.scrapy.scrapystreaming.Spider;
import org.scrapy.scrapystreaming.core.Callback;
import org.scrapy.scrapystreaming.Request;
import org.scrapy.scrapystreaming.messages.ResponseMessage;
import org.apache.commons.codec.binary.Base64;
import java.io.FileOutputStream;
import java.io.File;


/**
 * This examples demonstrate how to download binary data using Scrapy Streaming.
 *
 * It downloads an image using base64, and then save it to a file.
 */
public class RequestImage extends Spider {

    RequestImage() {
        name = "image";
    }

    public void parse(ResponseMessage response) {

    }

    public static void main(String args[]) throws Exception {

        RequestImage spider = new RequestImage();
        spider.start();

        // open a request to download the image
        Request r = new Request("http://httpbin.org/image/png");
        // set the body encoding to base64, so it can download the image using json
        r.base64 = true;

        r.open(new Callback() {
            public void onResponse(ResponseMessage response) {
                try {
                    // write the response body to a file and close the spider
                    byte data[] = Base64.decodeBase64(response.body);
                    FileOutputStream img = new FileOutputStream(new File("outputs/image.png"));
                    img.write(data);
                    img.close();
                    spider.close();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

    }

}
